plugins/InspectionGadgets/InspectionGadgetsAnalysis/src/com/siyeh/ig/bugs/IncorrectDateTimeFormatInspection.java

```diff
@@ -0,0 +1,262 @@
+import com.siyeh.InspectionGadgetsBundle;
+import com.siyeh.ig.callMatcher.CallMatcher;
+import com.siyeh.ig.psiutils.ExpressionUtils;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.*;
+import java.util.stream.Collectors;
+
+import static com.siyeh.ig.callMatcher.CallMatcher.*;
+
+public class IncorrectDateTimeFormatInspection extends AbstractBaseJavaLocalInspectionTool {
+
+  private interface CountVerifier {
+    boolean verifier(int count);
+  }
+
+  private static class RangeVerifier implements CountVerifier {
+    private final int from;
+    private final int to;
+
+    private RangeVerifier(int from, int to) {
+      this.from = from;
+      this.to = to;
+    }
+
+    @Override
+    public boolean verifier(int count) {
+      return count >= from && count <= to;
+    }
+  }
+
+  private static RangeVerifier rangeOf(int from, int to) {
+    return new RangeVerifier(from, to);
+  }
+
+  private static class SetVerifier implements CountVerifier {
+    private final Set<Integer> set;
+
+    private SetVerifier(int... numbers) {
+      set = Arrays.stream(numbers).boxed()
+        .collect(Collectors.toSet());
+    }
+
+    @Override
+    public boolean verifier(int count) {
+      return set.contains(count);
+    }
+  }
+
+  private static SetVerifier setOf(int... numbers) {
+    return new SetVerifier(numbers);
+  }
+
+  Map<Character, CountVerifier> DATE_TIME_FORMATTER_ALLOWED = Map.ofEntries(
+    Map.entry('G', rangeOf(1, 5)),
+    Map.entry('u', rangeOf(1, 19)),
+    Map.entry('y', rangeOf(1, 19)),
+    Map.entry('Y', rangeOf(1, Integer.MAX_VALUE)), //according to refs must be until 19, but now there is no exception
+    Map.entry('Q', rangeOf(1, 5)),
+    Map.entry('q', rangeOf(1, 5)),
+    Map.entry('M', rangeOf(1, 5)),
+    Map.entry('L', rangeOf(1, 5)),
+    Map.entry('w', rangeOf(1, 2)),
+    Map.entry('W', rangeOf(1, 1)),
+    Map.entry('d', rangeOf(1, 2)),
+    Map.entry('D', rangeOf(1, 3)),
+    Map.entry('F', rangeOf(1, 1)),
+    Map.entry('g', rangeOf(1, 19)),
+    Map.entry('E', rangeOf(1, 5)),
+    Map.entry('e', rangeOf(1, 5)),
+    Map.entry('c', rangeOf(1, 5)),
+
+    Map.entry('a', rangeOf(1, 1)),
+    Map.entry('h', rangeOf(1, 2)),
+    Map.entry('H', rangeOf(1, 2)),
+    Map.entry('k', rangeOf(1, 2)),
+    Map.entry('K', rangeOf(1, 2)),
+    Map.entry('m', rangeOf(1, 2)),
+    Map.entry('s', rangeOf(1, 2)),
+    Map.entry('S', rangeOf(1, 9)), //limit as nanos
+    Map.entry('A', rangeOf(1, 19)),
+    Map.entry('n', rangeOf(1, 19)),
+    Map.entry('N', rangeOf(1, 19)),
+
+    Map.entry('B', setOf(1, 4, 5)),
+
+    Map.entry('V', rangeOf(2, 2)),
+    Map.entry('v', setOf(1, 4)),
+    Map.entry('z', rangeOf(1, 4)),
+    Map.entry('O', setOf(1, 4)),
+    Map.entry('X', rangeOf(1, 5)),
+    Map.entry('x', rangeOf(1, 5)),
+    Map.entry('Z', rangeOf(1, 5))
+  );
+
+  private static final CallMatcher PATTERN_METHODS = anyOf(
```

Comment:
com.siyeh.ig.bugs.SuspiciousDateFormatInspection matches also SimpleDateFormat.applyPattern/applyLocalizedPattern which are (hopefully) using the same syntax. Also, a constructor for SimpleDateFormat (via `visitNewExpression`). Probably it's possible to reuse some logic from there?

Reply:
This is a great question. I've thought about that, but there are several problems, which complicate this inspection
The main problem is that the syntax patterns are similar, when common cases are used, but they have completely different limits.
For example:
   - `DateTimeFormatter.ofPattern("pppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppMMpM");` 
we don't expect any exception.
But ` new SimpleDateFormat("pppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppppMMpM")` produces exception, as 'p' is unsupported token.

- Patterns for DateTimeFormatter have many limits (like "MMMMMM" is too big). But SimpleDateFormat doesn' have these limits ("X" is an exception, and must be less than 4 - java.text.SimpleDateFormat#encode)

- DateTimeFormatter doesn't use optional brackets.

About `applyLocalizedPattern`, it seems to be too complex to check, because 
```
    public void applyLocalizedPattern(String pattern) {
         String p = translatePattern(pattern,
                                     formatData.getLocalPatternChars(),
                                     DateFormatSymbols.patternChars);
         compiledPattern = compile(p);
         this.pattern = p;
    }
```
and  it depends on local patterns, which can contain absolutely different tokens. 
Here, I replace standard token for year "y" to "a";
```
import java.text.DateFormatSymbols;
import java.text.SimpleDateFormat;
import java.util.Date;

public class Test {
    public static void main(String[] args) {
        SimpleDateFormat p = new SimpleDateFormat();
        DateFormatSymbols newFormatSymbols = new DateFormatSymbols();
        newFormatSymbols.setLocalPatternChars("Ga"); //standard pattern - "GyMdkHmsSEDFwWahKzZYuXL". I change the second letter.
        p.setDateFormatSymbols(newFormatSymbols);
        p.applyLocalizedPattern("aaaa");
        System.out.println(p.format(new Date())); //2022 if we live in 2022 year.
    }
}
```

Honestly, I am not sure, but, perhaps, `applyLocalizedPattern` doesn't suit for `SuspiciousDateFormatInspection` because of that.

Reply:
Ok, fair enough. Let's keep it as is for now.